### PR TITLE
Add option for user to download their data; #40

### LINKED ISSUE #40: Option to download user submission data
Though unlikely, it is possible that we may at any time lose portions of our database to unforeseen bugs. Or that we may be asked by regional contest organizers to remove their problems from the DMOJ if/when they develop their own OJ (funding et al.)

When this happens, it would be highly inconvenient to those who have submitted those problems to have to visit each problem and save their source. In general, a backup feature located on the users page would be useful in the long run.

It would operate with two options: `Backup all` and `Backup best`. Based on the option, a zip/tar file would be generated and the user would be able to download it. Very convenient!

This feature could be extended to include a `Download` option on submission pages.

### PR DESCRIPTION
**Pull request by Ninjaclasher**
Closes #40.

### CODE REVIEW

**Comment by Ninjaclasher** [comment]
image
image
image

**Comment by Ninjaclasher** [comment]
Docs for setting up the data cache location will come soon.

File: dmoj/settings.py [hunk 1]
```diff
@@ -72,6 +73,10 @@
 DMOJ_BLOG_RECENTLY_ATTEMPTED_PROBLEMS_COUNT = 7
 DMOJ_TOTP_TOLERANCE_HALF_MINUTES = 1
 DMOJ_USER_MAX_ORGANIZATION_COUNT = 3
+# Whether to allow users to download their data
+DMOJ_USER_DATA_DOWNLOAD = True
+DMOJ_USER_DATA_CACHE = ''
+DMOJ_USER_DATA_RATELIMIT = datetime.timedelta(days=1)
```
Comment (Xyene) [inline dmoj/settings.py:79]:
```suggestion
DMOJ_USER_DATA_DOWNLOAD_RATELIMIT = datetime.timedelta(days=1)
```

File: judge/forms.py [hunk 2]
```diff
@@ -67,6 +67,36 @@ def __init__(self, *args, **kwargs):
...
 
 
+class DownloadDataForm(Form):
+    comment_download = BooleanField(required=False, label=_('Download comments?'))
+    submission_download = BooleanField(required=False, label=_('Download submissions?'))
+    submission_problem_regex = CharField(initial='.*', label=_('Filter by problem code regex:'))
+    submission_results = MultipleChoiceField(
+        required=False,
+        widget=Select2MultipleWidget(
+            attrs={'style': 'width: 200px', 'data-placeholder': _('Leave empty to not filter')},
```
Comment (Xyene) [inline judge/forms.py:77]:
```suggestion
            attrs={'style': 'width: 200px', 'data-placeholder': _('Leave empty to include all submissions')},
```

File: judge/forms.py [hunk 3]
```diff
@@ -67,6 +67,36 @@ def __init__(self, *args, **kwargs):
...
+        ),
+        choices=sorted(map(itemgetter(0, 0), Submission.RESULT)),
+        label=_('Filter by result:'),
+    )
+
+    def clean(self):
+        can_download = ('comment_download', 'submission_download')
+        if not any(self.cleaned_data[v] for v in can_download):
+            raise ValidationError(_('Cannot download an empty zip! Select at least one thing to download.'))
```
Comment (Xyene) [inline judge/forms.py:86]:
```suggestion
            raise ValidationError(_('Please select at least one thing to download.'))
```

File: judge/views/user.py [hunk 4]
```diff
@@ -227,6 +233,94 @@ def get(self, request, *args, **kwargs):
...
+class UserPrepareData(LoginRequiredMixin, UserDataMixin, TitleMixin, FormView):
+    template_name = 'user/prepare-data.html'
+    form_class = DownloadDataForm
+
+    @cached_property
+    def _now(self):
+        return timezone.now()
+
+    @cached_property
+    def can_reprepare_data(self):
```
Comment (Xyene) [inline judge/views/user.py:256]: weird naming here

File: judge/views/user.py [hunk 5]
```diff
@@ -227,6 +233,94 @@ def get(self, request, *args, **kwargs):
...
+        return super().post(request, *args, **kwargs)
+
+
+class UserDownloadData(LoginRequiredMixin, UserDataMixin, View):
+    def get(self, request, *args, **kwargs):
+        if not os.path.exists(self.data_path):
+            raise Http404()
+
+        response = HttpResponse()
+        if hasattr(settings, 'DMOJ_USER_DATA_INTERNAL') and \
```
Comment (Xyene) [inline judge/views/user.py:313]: We have this `X-Accel-Redirect` hack in at least two places now; its complicated enough that it ought to be in a util file or something.

File: templates/user/prepare-data.html [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        $(document).ready(function () {
+            $('.time-remaining').each(function () {
+                count_down($(this));
+            });
+        });
+    </script>
+{% endblock %}
+
+{% block body %}
+    <div class="center-float">
```
Comment (Xyene) [inline templates/user/prepare-data.html:54]: Going by the screenshots, I think a bit of padding above this would be nice.

**Comment by Ninjaclasher** [comment]
All requested changes have been commited.

**Comment by Ninjaclasher** [comment]
This PR had a similar issue to #1055, which has been fixed in the latest commit.

**Comment by Ninjaclasher** [comment]
It seems that there are still a few minor issues that need to be ironed out before this can be merged.

**Comment by Ninjaclasher** [comment]
`regex` -> `glob`

File: judge/tasks/user.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        if result is None:
+            result = problem_code_glob_cache[code] = fnmatch.fnmatch(code, options['submission_problem_glob'])
+        return result
+
+    return list(filter(problem_code_glob_match, queryset))
+
+
+def apply_comment_filter(queryset, options):
+    if not options['comment_download']:
+        return list(queryset.none())
```
Comment (Xyene) [inline judge/tasks/user.py:37]: Is this different from `[]`?

File: templates/user/prepare-data.html [hunk 8]
```diff
@@ -0,0 +1,133 @@
+{% extends "base.html" %}
+
+{% block media %}
+    {{ form.media.css }}
+    <style>
+        .centered-form {
+            padding-top: 15px;
+            max-width: 460px;
+        }
+        .form-area {
```
Comment (Xyene) [inline templates/user/prepare-data.html:10]: Please add a single blank newline after each of these.

File: templates/user/prepare-data.html [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            }
+        }
+    </style>
+{% endblock %}
+
+{% block js_media %}
+    {{ form.media.js }}
+    <script type="text/javascript">
+        $(function () {
+            function check($obj, duration) {
```
Comment (Xyene) [inline templates/user/prepare-data.html:45]:
```suggestion
            function check ($obj, duration) {
```

File: judge/tasks/user.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                        'date': submission.date.isoformat(),
+                        'time': submission.time,
+                        'memory': submission.memory,
+                        'language': submission.language.key,
+                        'status': submission.status,
+                        'result': submission.result,
+                        'case_points': submission.case_points,
+                        'case_total': submission.case_total,
+                    }
+                    with data_file.open('submissions/%s.txt' % submission.id, 'w') as f:
```
Comment (Xyene) [inline judge/tasks/user.py:65]: Minor usability thing here... we can pull a bit of a hack to give these extensions:

https://github.com/DMOJ/judge-server/blob/3f60681c198df9fb3ae4ab8f2d0d4b8af7690535/dmoj/commands/submit.py#L59-L61

Comment (Ninjaclasher) [reply judge/tasks/user.py]: Added a commit to give submissions extensions.

**Comment by Xyene** [comment]
Thanks!

**Pull request merged**